webapp/src/assetEditor.tsx

```diff
@@ -14,39 +17,126 @@ function init() {
     const assetDiv = document.getElementById("asset-editor-field-div") as HTMLDivElement;
     ReactDOM.render(<AssetEditor />, assetDiv);
 }
+
+type AssetType = "sprite" | "tilemap";
+
+interface AssetEditorState {
+    viewType: AssetType;
+}
+
 export interface Message {
     data: MessageData;
 }
 export interface MessageData {
     type: string;
     message?: string;
     _fromVscode?: boolean;
+    name?: string;
 }
 
-export class AssetEditor extends React.Component {
+const DEFAULT_NAME = "tilemap_asset";
+
+export class AssetEditor extends React.Component<{}, AssetEditorState> {
     private editor: FieldEditorComponent<any>;
+    protected tilemapProject: pxt.TilemapProject;
+    protected tilemapName: string;
+
+    constructor(props: {}) {
+        super(props);
+
+        let view: AssetType = "sprite";
+        const v = /view(?:[:=])([a-zA-Z]+)/i.exec(window.location.href)
+        if (v && v[1] === "tilemap") {
+            view = "tilemap";
+        }
+        this.state = { viewType: view };
+
+        setTelemetryFunction(tickAssetEditorEvent);
+    }
 
     handleMessage = (msg: Message)  => {
         if (msg.data._fromVscode) {
             if (msg.data.type === "initialize") {
-                this.editor.loadJres(msg.data.message);
+                switch (this.state.viewType) {
+                    case "sprite":
+                        this.editor.loadJres(msg.data.message);
+                        break;
+                    case "tilemap":
+                        this.tilemapName = msg.data.name || DEFAULT_NAME;
+                        this.initTilemap(msg.data.message);
+                        break;
+                }
             } else if (msg.data.type === "update") {
                 this.sendJres();
             }
         }
     }
 
+    postMessage(msgData: MessageData) {
+        window.parent.postMessage(msgData, "*");
+    }
+
     refHandler = (e: FieldEditorComponent<any>) => {
         this.editor = e;
+        if (this.state.viewType === "tilemap") {
+            this.initTilemap();
+        }
     }
 
-    componentDidMount() {
-        window.addEventListener("message", this.handleMessage, null);
-        this.postMessage({type: "ready"});
+    initTilemap(s?: string) {
+        this.tilemapProject = new pxt.TilemapProject();
+        this.tilemapProject.loadJRes(s ? this.parseJRes(s) : {});
+        let project = this.tilemapProject.getTilemap(this.tilemapName);
+
+        if (!project) {
+            const [ name, map ] = this.tilemapProject.createNewTilemap(this.tilemapName, 16, 16, 16);
```

Comment:
we should think about ways to pass in a tile width

Reply:
yeah probably some UI on the vscode side for tile width + optional initial tilemap width/height, for the extension. i moved the tile width out so it can be set with that first 'initialize' message to the iframe